client/src/common/client.ts

```diff
@@ -1196,6 +1210,95 @@ class WillSaveWaitUntilFeature implements DynamicFeature<TextDocumentRegistratio
 	}
 }
 
+class WillCreateFilesFeature implements DynamicFeature<FileOperationRegistrationOptions> {
+
+	private _listener: Disposable | undefined;
+	private _globPatterns: Map<string, RegExp> = new Map<string, RegExp>();
+
+	constructor(private _client: BaseLanguageClient) {
+	}
+
+	public get registrationType(): RegistrationType<FileOperationRegistrationOptions> {
+		return WillCreateFilesRequest.type;
+	}
+
+	public fillClientCapabilities(capabilities: ClientCapabilities): void {
+		let value = ensure(capabilities, 'files')!;
+		value.willCreate = true;
+	}
+
+	public initialize(capabilities: ServerCapabilities, _: DocumentSelector): void {
+		let syncOptions = capabilities.files;
+		if (syncOptions?.willCreate?.globPattern) {
```

Comment:
Regarding errors: during initialization we should check the pattern, and if not valid ignore the capability and add something to the output channel.

If an invalid pattern is received during dynamic registration we should throw. However we should always throw `new Error` and never a string.

Reply:
Done both of these. We throw a custom error type in registration so it can be caught by static registration and ignored/logged instead. Let me know if this should done differently!